edb/edgeql/compiler/stmt.py

```diff
@@ -181,14 +181,19 @@ def compile_ForQuery(
                                 context=iterator.context)
 
         # Compile the body
-        stmt.result = compile_result_clause(
-            qlstmt.result,
-            view_scls=ctx.view_scls,
-            view_rptr=ctx.view_rptr,
-            result_alias=qlstmt.result_alias,
-            view_name=ctx.toplevel_result_view_name,
-            forward_rptr=True,
-            ctx=sctx)
+        with sctx.newscope(fenced=True) as bctx:
+            stmt.result = setgen.scoped_set(
+                compile_result_clause(
+                    qlstmt.result,
```

Comment:
`astutils.ensure_qlstmt(qlstmt.result)` here should be more efficient as it wouldn't introduce extraneous fences for expressions that are already statements.

Reply:
More efficient in what sense? Compiler efficiency, or will the generated code be improved?

`ensure_qlstmt` breaks a few tests---I haven't dug into it yet though.

Reply:
I was mainly concerned with scope tree complexity, but the benefit here is negligible.  Let's keep your initial version then.  Technically, `ensure_qlstmt` should be equivalent to setting up a fence, but there's probably some weird interaction with iterator state.